core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20250428_entity_keys_update_trigger.xml

```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog logicalFilePath="dummy" xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.31.xsd">
+
+
+    <!-- update the ENTITY triggers which populate ENTITY_KEYS to reflect the latest ENTITY.attributes format -->
+    <changeSet logicalFilePath="dummy" author="davidan" id="update_entity_keys_triggers_for_migration">
+        <sql stripComments="true" endDelimiter="~">
+            DROP TRIGGER IF EXISTS after_entity_update ~
+
+            CREATE TRIGGER after_entity_update AFTER UPDATE ON ENTITY
+                FOR EACH ROW
+                BEGIN
+                    declare new_keys json;
+                    declare old_keys json;
+                    -- is this row soft-deleted?
+                    if old.deleted = 0 and new.deleted = 1 then
+                        DELETE FROM ENTITY_KEYS WHERE id = new.id;
+                    elseif new.attributes is not null then
+                        -- compare old keys to new keys; update the ENTITY_KEYS table only if they are different
+                        set new_keys := ifnull(JSON_KEYS(new.attributes, '$.attrs'), '[]');
+                        set old_keys := ifnull(JSON_KEYS(old.attributes, '$.attrs'), '[]');
+                        if JSON_LENGTH(new_keys) != JSON_LENGTH(old_keys) OR JSON_CONTAINS(new_keys, old_keys) = 0 then
+                            -- upsert into ENTITY_KEYS. The row should exist except when migrating from legacy data tables.
+                            insert into ENTITY_KEYS (id, workspace_id, entity_type, attribute_keys, last_updated)
+                            values (new.id, new.workspace_id, new.entity_type, new_keys, now(3))
+                            on duplicate key update attribute_keys = new_keys, last_updated = now(3);
```

Comment:
The main change is here; previously this was an `update` statement, now it is an `insert … on duplicate key update` statement. This is necessary for the migration which updates entities in place; those entities do not already have a row in `ENTITY_KEYS`.